watcher.go

```diff
@@ -0,0 +1,128 @@
+	MaxAttempts int
+}
+
+var (
+	DefaultWatcher = &Watcher{}
+
+	// WatchTransport is the same as DefaultTransport with a longer
+	// ResponseHeaderTimeout.  This is copied from DefaultTransport.  We don't
+	// do an actual copy because the Trasnport uses mutexes.  copied mutexes
+	// are ineffective.  You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport for watch.
+	WatchTransport http.RoundTripper = &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout: 5 * time.Second,
+		}).DialContext,
+		DisableCompression:    true,
+		MaxIdleConns:          5,
+		MaxIdleConnsPerHost:   2,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   5 * time.Second,
+		ResponseHeaderTimeout: 1 * time.Minute,
+		ExpectContinueTimeout: 5 * time.Second,
+	}
+)
+
+func (w *Watcher) client() *Client {
+	if client := w.Client; client != nil {
+		return client
+	}
+	c := &Client{
+		Transport: WatchTransport,
+	}
+	return c
+}
+
+// Watch is the package-level Watch defintion which is called on
+// DefaultWatcher.
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+// WatchPrefix is the package-level WatchPrefix defintion which is called on
+// DefaultWatcher.
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+// Watch executes a long poll for changes to the given key.  handler will be
+// called immediately upon registration to initialize the watch and returns
+// the initial value (as a list, this is what Consul API returns).  There is
+// a chance that Watch can miss updates due to the way the API is implemented.
+// See https://github.com/hashicorp/consul/issues/1761 for details.
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+// WatchPrefix executes a long poll for changes to anything under the given
+// prefix.  handler will be called immediately upon registration to initialize
+// the watch and returns the initial value (as a list, this is what Consul API
+// returns).  There is a chance that WatchPrefix can miss updates due to the
+// way the API is implemented.  See https://github.com/hashicorp/consul/issues/1761
+// for details.
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	if w.MaxAttempts <= 0 {
+		w.MaxAttempts = 10
+	}
+
+	path := "/v1/kv/" + key
+	value := "0"
+
+	attempt := 0
+	for {
+		q.Add(Param{Name: "index", Value: value})
+		resp := []KeyData{}
+		hdr, err := w.client().do(ctx, "GET", path, q, nil, &resp)
+		if err != nil {
+			attempt += 1
+			if attempt >= w.MaxAttempts {
+				handler(nil, err)
+				return
```

Comment:
Do you still feel like we should abort here?

Reply:
Thinking about this overnight, I think we should remove this code.  You're right.

I'll put it in the docs that the caller is responsible for backoffs and cancelling the context if they so choose.

I think this means we can remove all the attempt counting code too.